modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/CSharpClientCodegen.java

```diff
@@ -156,133 +173,160 @@ public CSharpClientCodegen() {
         }
         if (isEmpty(modelPackage)) {
-            modelPackage = "Model";
+            setModelPackage("Model");
         }
         clientPackage = "Client";
 
         Boolean excludeTests = false;
         if (additionalProperties.containsKey(CodegenConstants.EXCLUDE_TESTS)) {
-            excludeTests = Boolean.valueOf(additionalProperties.get(CodegenConstants.EXCLUDE_TESTS).toString());
+            excludeTests = convertPropertyToBooleanAndWriteBack(CodegenConstants.EXCLUDE_TESTS);
         }
 
-        additionalProperties.put(CodegenConstants.API_PACKAGE, apiPackage);
-        additionalProperties.put(CodegenConstants.MODEL_PACKAGE, modelPackage);
-        additionalProperties.put("clientPackage", clientPackage);
-        additionalProperties.put("emitDefaultValue", optionalEmitDefaultValue);
-
-        if (!additionalProperties.containsKey("validatable")) {
-            // default validatable to true if not set
-            additionalProperties.put("validatable", true);
+        if (additionalProperties.containsKey(CodegenConstants.VALIDATABLE)) {
+            setValidatable(convertPropertyToBooleanAndWriteBack(CodegenConstants.VALIDATABLE));
+        } else {
+            additionalProperties.put(CodegenConstants.VALIDATABLE, validatable);
         }
 
         if (additionalProperties.containsKey(CodegenConstants.DOTNET_FRAMEWORK)) {
             setTargetFramework((String) additionalProperties.get(CodegenConstants.DOTNET_FRAMEWORK));
         } else {
             // Ensure default is set.
             setTargetFramework(NET45);
-            additionalProperties.put("targetFramework", this.targetFramework);
+            additionalProperties.put(CodegenConstants.DOTNET_FRAMEWORK, this.targetFramework);
         }
 
         if (NET35.equals(this.targetFramework)) {
+            // This is correct, mono will require you build .NET 3.5 sources using 4.0 SDK
+            additionalProperties.put(MCS_NET_VERSION_KEY, "4");
+            additionalProperties.put("net35", true);
+            if (additionalProperties.containsKey(CodegenConstants.SUPPORTS_ASYNC)) {
+                LOGGER.warn(".NET 3.5 generator does not support async.");
+                additionalProperties.remove(CodegenConstants.SUPPORTS_ASYNC);
+            }
+
             setTargetFrameworkNuget("net35");
+            setValidatable(Boolean.FALSE);
             setSupportsAsync(Boolean.FALSE);
-            if (additionalProperties.containsKey("supportsAsync")) {
-                additionalProperties.remove("supportsAsync");
-            }
-            additionalProperties.put("validatable", false);
-            additionalProperties.put("net35", true);
         } else if (NETSTANDARD.equals(this.targetFramework)) {
+            // TODO: NETSTANDARD here is misrepresenting a PCL v5.0 which supports .NET Framework 4.6+, .NET Core 1.0, and Windows Universal 10.0
+            additionalProperties.put(MCS_NET_VERSION_KEY, "4.6-api");
```

Comment:
NOTE: I added this, although this `NETSTANDARD` isn't writing `build.sh` at the moment.